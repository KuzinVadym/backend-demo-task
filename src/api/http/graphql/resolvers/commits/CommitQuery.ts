import { Arg, Ctx, Int, Query, Resolver } from 'type-graphql';
import { withHttpService } from '../../../shared/utils/withHttpService';
import { CommitType } from './CommitType';

@Resolver((of) => CommitType)
export class CommitQuery {
  @Query(() => [CommitType])
  public async commits(@Ctx() ctx): Promise<CommitType[]> {
    const userHttpService = withHttpService('commits', ctx.getState);
    return userHttpService.getCommits();
  }

  @Query(() => CommitType)
  public async commit(
    @Arg('id', (type) => Int) id: number,
    @Ctx() ctx
  ): Promise<CommitType> {
    const userHttpService = withHttpService('commits', ctx.getState);
    return userHttpService.getCommit(id);
  }
}
