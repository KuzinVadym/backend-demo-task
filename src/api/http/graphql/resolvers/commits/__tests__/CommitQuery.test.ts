import 'reflect-metadata';
import {
  ApolloServerTestClient,
  createTestClient
} from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server-express';
import { buildSchema } from 'type-graphql';
import { CommitQuery } from '../CommitQuery';

const contextMock = {
  getState: jest.fn()
};

const commitsServiceMock = {
  getCommits: jest.fn(),
  getCommit: jest.fn()
};

describe('CommitQuery', () => {
  let server: ApolloServer;

  beforeAll(async () => {
    const schema = await buildSchema({ resolvers: [CommitQuery] });
    server = new ApolloServer({
      schema,
      context: contextMock
    });
  });

  beforeEach(async () => {
    commitsServiceMock.getCommits.mockResolvedValue([
      {
        userId: 1,
        taskId: 1,
        status: 'ToDo',
        createdAt: new Date()
      }
    ]);
    commitsServiceMock.getCommit.mockResolvedValue({
      userId: 1,
      taskId: 1,
      status: 'ToDo',
      createdAt: new Date()
    });

    contextMock.getState.mockReturnValue({
      httpServices: {
        commits: commitsServiceMock
      }
    });
  });

  it('returns tasks data', async () => {
    const testQuery = `{
      commits {
        taskId
        userId
        status
      }
    }
  `;

    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: {} });
    expect(res).toMatchSnapshot();
  });

  it('returns task data', async () => {
    const testQuery = `
    query CommitQuery($id: Int!){
      commit(id: $id) {
        taskId
        userId
        status
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    const res = await query({ query: testQuery, variables: { id: 1 } });
    expect(res).toMatchSnapshot();
  });

  it('cals service with right params', async () => {
    const testQuery = `
    query CommitQuery($id: Int!){
      commit(id: $id) {
        taskId
        userId
        status
      }
    }
  `;
    const { query }: ApolloServerTestClient = createTestClient(server);

    await query({ query: testQuery, variables: { id: 1 } });

    expect(commitsServiceMock.getCommit).toBeCalledWith(1);
  });
});
