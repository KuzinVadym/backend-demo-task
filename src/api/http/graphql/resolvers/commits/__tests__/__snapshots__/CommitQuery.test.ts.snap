// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitQuery returns task data 1`] = `
Object {
  "data": Object {
    "commit": Object {
      "status": "ToDo",
      "taskId": 1,
      "userId": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`CommitQuery returns tasks data 1`] = `
Object {
  "data": Object {
    "commits": Array [
      Object {
        "status": "ToDo",
        "taskId": 1,
        "userId": 1,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
